test/e2e/activator_test.go

```diff
@@ -0,0 +1,193 @@
+func TestActivatorOverload(t *testing.T) {
+	var (
+		clients *test.Clients
+		logger  *logging.BaseLogger
+	)
+	const (
+		// The number of concurrent requests to hit the activator with.
+		// 1000 = the number concurrent connections in Istio.
+		concurrency = 1000
+		// Timeout to wait for the responses.
+		// Ideally we must wait ~30 seconds, TODO: need to figure out where the delta comes from.
+		timeout = 65 * time.Second
+		// How long the service will process the request in ms.
+		serviceSleep = 300
+	)
+	logger = logging.GetContextLogger(t.Name())
+	clients = Setup(t)
+
+	helloWorldNames := test.ResourceNames{
+		Service: test.AppendRandomString(configName, logger),
+		Image:   "observed-concurrency",
+	}
+
+	configOptions := test.Options{
+		ContainerConcurrency: 1,
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = make(map[string]string)
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations["autoscaling.knative.dev/maxScale"] = "10"
+	}
+
+	test.CleanupOnInterrupt(func() { TearDown(clients, helloWorldNames, logger) }, logger)
+	defer TearDown(clients, helloWorldNames, logger)
+
+	// Create a service with concurrency 1 that could sleep for N ms.
+	// Limit its maxScale to 10 containers, wait for the service to scale down and hit it with concurrent requests.
+	resources, err := test.CreateRunLatestServiceReady(logger, clients, &helloWorldNames, &configOptions, fopt)
+	if err != nil {
+		t.Fatalf("Unable to create resources: %v", err)
+	}
+	domain := resources.Route.Status.Domain
+
+	deploymentName := resources.Revision.Name + "-deployment"
+
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	endpoint, err := spoof.GetServiceEndpoint(clients.KubeClient.Kube)
+	if err != nil {
+		t.Fatalf("Could not get service endpoint for spoofing client")
+	}
+
+	url := fmt.Sprintf("http://%s/?timeout=%d", *endpoint, serviceSleep)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, logger, domain, test.ServingFlags.ResolvableDomain)
+
+	responseChannel := make(chan *spoof.Response, concurrency)
+	roundTrip := roundTrip(client, url)
+
+	sendRequests(roundTrip, concurrency, responseChannel, timeout, logger, t)
+
+	analyseResponses(responseChannel, concurrency, timeout, t)
+}
+
+func sendRequests(roundTrip func() (*spoof.Response, error), concurrency int, resChannel chan *spoof.Response, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) {
+	var (
+		group     errgroup.Group
+		responses int32
+	)
+	doneChan := make(chan struct{})
+	timeoutChan := time.After(timeout)
```

Comment:
Do you want this channel to be created here or above the loop below?

Reply:
I'm fine with leaving it where it is now to have all channels defined in one place, for the sake of readability. I think it adding some microseconds should not be that critical for this test. However it could arguable.